### DISCUSSION TITLE
What about iOS support?

### DISCUSSION THREAD

**Discussion by agnickolov** [paullouisageneau/libdatachannel#433] [Ideas]
Are there plans to add build support for iOS? I imagine it shouldn't be hard to port over the OS X build to iOS.

**Comment by paullouisageneau** [comment]
Actually, the library should compile for iOS, only I don't have Apple devices to test. Someone managed to compile an earlier version so I'm confident it still works (if there are incompatibilities, they should be minimal).

Keep me updated if you try compiling for iOS.

**Reply by agnickolov** [reply]
It took me a while to get CMake to generate an iOS Xcode project. I did some invasive surgery on CMakeLists.txt and I had to change some system CMake variables to enable it to find dependencies when cross-compiling for iOS. With that out of the way, the next roadblock is that usrsctp does not compile for iOS (ARM64 architecture). I'll keep plugging at it.

**Reply by agnickolov** [reply]
I investigated on the usrsctp project and it does not advertise support for iOS. I ran into numerous issues trying to compile it for iOS too. I suspect my remaining option is going to be Google's complete WebRTC library. They are moving to their own implementation called dcsctp. Fingers crossed it supports iOS...

**Reply by paullouisageneau** [reply]
> With that out of the way, the next roadblock is that usrsctp does not compile for iOS (ARM64 architecture).

I'm surprised, as usrsctp compiles fine for Android ARM64 and Raspberry Pi (armv7l). Is it a specific iOS issue?  

> I investigated on the usrsctp project and it does not advertise support for iOS. I ran into numerous issues trying to compile it for iOS too. I suspect my remaining option is going to be Google's complete WebRTC library. They are moving to their own implementation called dcsctp. Fingers crossed it supports iOS...

Before moving to their own implementation, Google's WebRTC has used usrsctp and it could be built for iOS.

I see there are fixes for iOS compilation [recently merged in usrsctp](https://github.com/sctplab/usrsctp/commit/e984d7f3c1b13d0b0582497b385c93f0e8d89fb3). Could you please try pulling the usrsctp submodule to last master to see if it fixes the problem?

**Reply by agnickolov** [reply]
That got me a bit further. I needed to edit 7 more files in order to get It to compile, however. Mostly a bunch of type casts, but also needed to define __APPLE_USE_RFC_2292 for some obsolete IPV6_PKTINFO constant. I'm using Xcode 12.4 and looks like the code has not been updated to handle it.

**Reply by agnickolov** [reply]
I cleaned up all CMakeLists edits for libdatachannel itself. The necessary changes for iOS boil down to the following line replicated for all executables:
	set_target_properties(datachannel-tests PROPERTIES
		XCODE_ATTRIBUTE_PRODUCT_BUNDLE_IDENTIFIER com.github.paullouisageneau.libdatachannel.tests)
Xcode needs a bundle identifier for every executable built for iOS.
There is one such change needed for libjuice as well.

Of note, the CMake project generation line for iOS is quite lengthy:
cmake -B build_ios -DUSE_GNUTLS=0 -DUSE_NICE=0 -DOPENSSL_ROOT_DIR=\<path-to-openssl\> -GXcode -DCMAKE_SYSTEM_NAME=iOS -DCMAKE_OSX_ARCHITECTURES=arm64 -DCMAKE_XCODE_ATTRIBUTE_DEVELOPMENT_TEAM=\<profile-id\> -DCMAKE_XCODE_ATTRIBUTE_ENABLE_BITCODE=NO -DCMAKE_FIND_ROOT_PATH_MODE_PACKAGE=BOTH -DCMAKE_FIND_ROOT_PATH_MODE_LIBRARY=BOTH -DCMAKE_FIND_ROOT_PATH_MODE_INCLUDE=BOTH

**Reply by dbotha** [reply]
Thought I'd jump in here too as I've just got things building for iOS myself. For anyone that might be interested these are the changes I made to usrsctp to get it building under my setup (they seem inline with what @agnickolov made): 

https://github.com/sctplab/usrsctp/commit/4e3c089173b351adfa605a2f16ac327300b7b489

@agnickolov Oddly I didn't need to touch libdatachannel's `CMakeLists.txt` at all to get it built into my app, but I've yet to test things on a device to ensure it actually works 🤷‍♂️

**Reply by paullouisageneau** [reply]
Great! @dbotha will you merge your changes upstream into usrsctp?

**Reply by agnickolov** [reply]
Sorry, I've been busy at work writing a prototype with it. I'll create a PR for usrsctp with my fixes for XCode 12 soon when I have a few spare minutes.
@dbotha The changes for the bundle ID attribute are necessary when building for a device.

**Reply by dbotha** [reply]
@agnickolov Are you using the `libdatachannel`'s WebSocket functionality at all? I've got the build on my device now but the TLS handshake always fails (& it's not my certificate, https://www.google.com also fails to negotiate the TLS handshake):

    ERROR [rtc::impl::TlsTransport::InfoCallback@433] TLS alert: unknown CA
    ERROR [rtc::impl::TlsTransport::runRecvLoop@423] TLS handshake failed

Perhaps it's down to something wrong with my build of OpenSSL... Did you encounter anything similar? 

@paullouisageneau Once I'm confident I've got things working (and if @agnickolov hasn't already) I'll submit a PR to usrsctp.

**Reply by paullouisageneau** [reply]
@dbotha The TLS unknown CA error probably comes from the OpenSSL build missing the system CA certificates location. You can bypass the problem by setting `disableTlsVerification` in `WebSocket::Configuration` (or `rtcWsConfiguration` for the C API), but in that case the server certificate will be trusted blindly.

**Reply by dbotha** [reply]
@paullouisageneau I was just about to ask if there was a means to disable this, and yes my investigation today leads me to the same conclusion - the build is missing the CA certificates. I'll figure out how to get those bundled in due course but will check if things work end to end first after disabling. Thanks so much for you help on this!

**Reply by agnickolov** [reply]
I submitted a PR for usrsctp. Next are the PRs for libjuice and libdatachannel.

**Reply by agnickolov** [reply]
PRs for libjuice and libdatachannel submitted as well. I'll leave updating the submodules for you.

**Reply by paullouisageneau** [reply]
@agnickolov Thank you! It seems there is a [problem](https://github.com/sctplab/usrsctp/pull/593/checks?check_run_id=2824372345) with the usrsctp fix, as `__APPLE_USE_RFC_2292` should not be redefined if already defined.

**Reply by agnickolov** [reply]
I made a fix about that. Thank you for pointing it out.

**Comment by paullouisageneau** [comment]
For reference, iOS support was added in https://github.com/paullouisageneau/libdatachannel/issues/444 and shipped in version 0.15.

**Reply by dbotha** [reply]
Thanks @paullouisageneau !
